deepchecks/recommender/ranking.py

```diff
@@ -7,705 +7,423 @@
 
-    return avg_self_information
+    # Calculate novelty and normalized novelty
+    novelty = np.mean(all_self_information)
+    normalized_novelty = novelty / top_information_mean
+    return novelty, normalized_novelty
 
-
-def normalized_distance_based_performance_measure(relevant_items: List, recommendation: List):
+def diversity_score(predictions: List, target_vectors: Union[List, np.array],
+                    top_n: int = 10) -> float:
     """
-    Calculates the Normalized Distance-based Performance Measure (NPDM) between two
-    ordered lists. Two matching orderings return 0.0 while two unmatched orderings returns 1.0.
-
-    Args:
-        relevant_items (List): List of items
-        recommendation (List): The predicted list of items
+    Compute the diversity score of a set of recommendations.
+    Parameters:
+    predictions (List): List of predicted item indices for each user.
+    target_vectors (Union[List, np.array]): List or array of item feature vectors.
+    top_n (int): The number of top recommended items to consider.
 
     Returns:
-        NDPM (float): Normalized Distance-based Performance Measure
-
-    Metric Definition:
-    Yao, Y. Y. "Measuring retrieval effectiveness based on user preference of documents."
-    Journal of the American Society for Information science 46.2 (1995): 133-145.
-
-    Definition from:
-    Shani, Guy, and Asela Gunawardana. "Evaluating recommendation systems."
-    Recommender systems handbook. Springer, Boston, MA, 2011. 257-297
+        float: Diversity score.
     """
-    assert set(relevant_items) == set(recommendation)
-
-    item_relevant_items_rank = {item: i + 1 for i, item in enumerate(dict.fromkeys(relevant_items))}
-    item_predicted_rank = {item: i + 1 for i, item in enumerate(dict.fromkeys(recommendation))}
-
-    items = set(relevant_items)
-
-    item_combinations = itertools.combinations(items, 2)
-
-    C_minus = 0
-    C_plus = 0
-    C_u = 0
-
-    for item1, item2 in item_combinations:
-        item1_relevant_items_rank = item_relevant_items_rank[item1]
-        item2_relevant_items_rank = item_relevant_items_rank[item2]
+    # If there are no predictions, return 0.
+    if len(predictions) == 0:
+        return 0.0
 
-        item1_pred_rank = item_predicted_rank[item1]
-        item2_pred_rank = item_predicted_rank[item2]
+    # Compute the cosine similarity matrix between the target vectors.
+    sim_matrix = cosine_similarity(target_vectors[:top_n])
 
-        C = np.sign(item1_pred_rank - item2_pred_rank) * np.sign(item1_relevant_items_rank - item2_relevant_items_rank)
+    # We only consider the upper or lower matrix triangle and ignore the diagonal as well.
+    sim_avg = sim_matrix[np.tril_indices(len(sim_matrix), -1)].mean()
 
-        C_u += C**2
+    # Return the diversity score.
+    return 1 - sim_avg
 
-        if C < 0:
-            C_minus += 1
-        else:
-            C_plus += 1
+def mca(x_categorical: Union[np.array, List[List]],
```

Comment:
What is this used for?

Reply:
The MCA (Multiple Correspondence Analysis) is a method used to convert categorical features into numerical features. It uses the chi-square statistic, commonly employed for testing independence between categorical variables in a contingency table (using relative frequencies, and kind of conditional probability).

Alternatively, we could use the "overlap" measure which is awidely used to compare data with categorical features. It is popular due to its simplicity. The measure assigns a value of 1 if two rows have the same category in a specific column, and 0 otherwise.

However, combining the diversity matrix for numerical features and the overlap matrix for categorical features can pose challenges. To avoid this issue, using MCA to convert categorical features into numerical features is a recommended approach. MCA is simple, robust, and eliminates the need to merge different matrices.

Reply:
seems legit